packages/cron/src/jobs/measureNftTimeToRetrievability.js

```diff
@@ -0,0 +1,215 @@
+ * @property {number}       contentLength
+ * @property {Date}         startTime
+ * @property {Milliseconds} duration
+ */
+
+/**
+ * @typedef {import('nft.storage/dist/src/token').TokenInput} TokenInput
+ */
+
+/**
+ * @template {TokenInput} T
+ * @typedef {import('nft.storage').TokenType<T>} TokenType
+ */
+
+/**
+ * @typedef {(token: TokenInput) => Promise<Pick<TokenType<TokenInput>, 'ipnft'>>} StoreFunction
+ */
+
+/**
+ * @returns {AsyncIterable<File>}
+ */
+export function TestImages(count = 1) {
+  return (async function* () {
+    while (count--) {
+      yield new File([Uint8Array.from([1, 2, 3])], 'image.png', {
+        type: 'image/png',
+      })
+    }
+  })()
+}
+
+/**
+ * @param  {...URL} urls
+ * @returns {AsyncIterable<Blob>}
+ */
+export function UrlImages(...urls) {
+  return (async function* () {
+    for (const url of urls) {
+      const resp = await fetch(url.toString())
+      yield resp.blob()
+    }
+  })()
+}
+
+/**
+ * Job that tests/measures steps
+ * * prepare a sample image
+ * * upload to nft.storage
+ * * retrieve image through ipfs gateway
+ * @template {import('nft.storage/dist/src/lib/interface').TokenInput} T
+ * @param {object} config
+ * @param {AsyncIterable<Blob>} config.images - images to upload/retrieve
+ * @param {StoreFunction} [config.store] - function to store nft
+ * @param {string} [config.url] - URL to nft.storage to measure
+ * @param {string} [config.metricsPushGateway] - Server to send metrics to. should be a https://github.com/prometheus/pushgateway
+ * @param {import('../lib/log.js').LogFunction} config.log - logger
+ * @param {object} secrets
+ * @param {string} secrets.nftStorageToken - API Token for nft.storage
+ * @param {string} [secrets.metricsPushGatewayBasicAuthUser] - authorization for metricsPushGateway
+ */
+export async function measureNftTimeToRetrievability(config, secrets) {
+  const start = {
+    type: 'start',
+    job: 'measureNftTimeToRetrievability',
+    config,
+  }
+  config.log('info', start)
+  for await (const image of config.images) {
+    const imageId = Number(new Date()).toString()
+    const nft = {
+      image: /** @type {Blob} */ (image), // use image Blob as `image` field
+      name: 'Example NFT',
+      description: 'Example NFT used by nft.storage cron nft-ttr',
+    }
+    const client = new NFTStorage({
+      token: secrets.nftStorageToken,
+    })
+    const storeStartedDate = new Date()
+    const storeStartedAt = now()
+    /** @type {StoreFunction} */
+    const store = config.store || ((nft) => client.store(nft))
+    const metadata = await store(nft)
+    const storeEndAt = now()
+    const storeLog = {
+      type: 'store',
+      image: imageId,
+      startTime: storeStartedDate,
+      duration: new Milliseconds(
+        storeEndAt.toNumber() - storeStartedAt.toNumber()
+      ),
+    }
+    config.log('info', storeLog)
+    const retrievalUrl = new URL(
+      `https://${metadata.ipnft}.ipfs.nftstorage.link/image/blob`
```

Comment:
eventually should test a handful of gateways, not just this one. perhaps as separate job invocations (e.g. using github actions matrix)

Reply:
https://github.com/nftstorage/nft.storage/pull/1945/commits/176ebeab8fd1a665010290404db039f1b8644661